api/api.go

```diff
@@ -82,33 +98,85 @@ func (api *API) ListenAndServe(port int) error {
 	return http.ListenAndServe(fmt.Sprintf(":%d", port), api.staticRouter)
 }
 
-// WithDBSession injects a session context into the request context of the handler.
+// WithDBSession injects a session context into the request context of the
+// handler. In case of a MongoDB WriteConflict error, the call is retried up to
+// DBTxnRetryCount times or until the request context expires.
 func (api *API) WithDBSession(h httprouter.Handle) httprouter.Handle {
 	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-		// Create a new db session
-		sess, err := api.staticDB.NewSession()
-		if err != nil {
-			api.WriteError(w, errors.AddContext(err, "failed to start a new mongo session"), http.StatusInternalServerError)
-			return
+		numRetriesLeft := DBTxnRetryCount
+		var body []byte
+		var err error
+		if req.Body != nil {
+			// Read the request's body and replace its Body io.ReadCloser with a
+			// new one based off the read data.
+			body, err = io.ReadAll(io.LimitReader(req.Body, LimitBodySizeLarge))
+			if err != nil {
+				api.WriteError(w, errors.AddContext(err, "failed to read body"), http.StatusBadRequest)
+				return
+			}
+			_ = req.Body.Close()
 		}
-		// Close session after the handler is done.
-		defer sess.EndSession(req.Context())
-
-		// Create session context.
-		sctx := mongo.NewSessionContext(req.Context(), sess)
 
-		// Get the special response writer.
-		mw, err := NewMongoWriter(w, sctx, api.staticLogger)
-		if err != nil {
-			api.WriteError(w, errors.AddContext(err, "failed to start a new transaction"), http.StatusInternalServerError)
-			return
+		// handleFn wraps a full execution of the handler, combined with a retry
+		// detection and counting. It also takes care of creating and cancelling
+		// Mongo sessions and transactions.
+		handleFn := func() (retry bool) {
```

Comment:
the named return variable is not necessary here right?

Reply:
we avoid named return variables in `skyd` at all cost

Reply:
Interesting. Why is that? I am not aware of any drawbacks to using named return values.
There is nothing said about those in our Engineering Guidelines.

Named return values are a nice and unobtrusive way to document a function.
Here's what Effective Go has to say about them: https://go.dev/doc/effective_go#named-results

Reply:
Well I like them too to be honest, but in `skyd` we avoid them when we can because it's often a foot gun and can lead to unexpected bugs 🤷‍♂️ , it's easy to miss a place where you overwrite it or don't set it when you want it to be set.

I personally am in your camp and like the documenting aspect of it, and the fact that you skip the var initialisation, but I'm just doing `skyd` reviews here, I don't really mind whether you update it or not.

Reply:
The way it is used here it is probably fine. But for future reference we usually try to avoid them, especially in longer methods. Unless they significantly simplify code. e.g. when you need to close something in an error case.

The reason is that for longer functions they can really harm readability and can lead to weird errors. e.g. we had a surprising number of issues where we'd miss assignments to named return vars or shadow them and then miss errors and such.

If it was just for the sake of documentation this could be pulled out into a separate function `WithRetry` that has its own docstring and an explanation of what the return value is. Which should probably happen anyway in a f/u since it's enough code.